### PR TITLE
[MPS] Improve memory usage and performance utilizing garbage collector and adaptive commit

### LINKED ISSUE #79283: OOM Issues despite plenty of resources on M1 Mac
### 🐛 Describe the bug

In a few different instances, I've run into an issue where during the loss.backward() suddenly the program will return "Killed". The first few iterations loss.backward() will be successful and then it just dies. I've tracked the disk usage as part of my troubleshooting and tons of memory remain before it fails. It only appears to be happening on my M1 Mac Mini. If I shrink the dimensions immensely such as making image_size = 100 or vocab_size = 1000, it's able to finish running (through with obviously degraded resutls). I've tried strategically calling gc.collect() with no success. 

I wish I had more helpful information, but as far as I can tell the problem is two fold. First part is that the M1 Mac GPU is not recognized ie. torch.cuda.is_available() returns false and "mps" isn't recognized as an available device type in torch version 1.11.0. This means "cpu" is the only working device option. Second part seems to be the freeing of the cache.

### Versions

Python Version: 3.9
Ubuntu Version: 21.04 (dev-hirsute)
PyTorch Version: 1.11.0
TorchVision Version: 0.12.0

### PR DESCRIPTION
**Pull request by razarmehr**
- Improve memory usage and performance utilizing garbage collector and adaptive commit
- Enable low watermark limit to detect memory pressure.
- Enable garbage collection and adaptive commit strategies when under memory pressure.
- More efficient resource management by splitting large heaps (instead of reusing oversized buffers for smaller allocation requests)
- Introduce Extra Large heaps to improve performance by avoiding numerous costly allocation of smaller heaps
- Fix purgeability when releasing the Metal heaps
- Fix the race condition when deferring the heap's size update

Fixes #79283

### CODE REVIEW

**Review by kulinseth — APPROVED** [review]
Looks fine.

**Review by malfet — APPROVED** [review]

**Comment by dfalbel** [comment]
@razarmehr Do you think we could still call the memory callbacks registered here:

https://github.com/pytorch/pytorch/blob/fc94a2115b31dfe7a0d8f28eb4f5ed532c4f0792/aten/src/ATen/mps/MPSAllocator.mm#L12

Looks like their usage has been removed in this PR.
